packages/core/src/CreateNodeAsTool.ts

```diff
@@ -1,296 +1,432 @@
+						current += 2;
+						continue;
+					}
+
+					if (char === quoteChar) {
+						inQuotes = false;
+						quoteChar = '';
+					}
+					argsString += char;
+				} else {
+					// Handle characters outside quotes
+					if (char === "'" || char === '"') {
+						inQuotes = true;
+						quoteChar = char;
+					} else if (char === '(') {
+						parenthesesCount++;
+					} else if (char === ')') {
+						parenthesesCount--;
+					}
+
+					// Only add characters if we're still inside the main parentheses
+					if (parenthesesCount > 0 || char !== ')') {
+						argsString += char;
+					}
+				}
+
+				current++;
+			}
+
+			// If parentheses are balanced, parse the arguments
+			if (parenthesesCount === 0) {
+				try {
+					const parsedArgs = this.parseArguments(argsString);
+					args.push(parsedArgs);
+				} catch (error) {
+					// If parsing fails, throw an ApplicationError with details
+					throw new NodeOperationError(
+						this.ctx.getNode(),
+						`Failed to parse $fromAI arguments: ${argsString}: ${error}`,
+					);
+				}
+			} else {
+				// Log an error if parentheses are unbalanced
+				throw new NodeOperationError(
+					this.ctx.getNode(),
+					`Unbalanced parentheses while parsing $fromAI call: ${str.slice(startIndex)}`,
+				);
 			}
 		}
+
+		return args;
 	}
 
-	return results;
-}
+	/**
+	 * Parses the arguments of a single $fromAI function call.
+	 * @param argsString The string containing the function arguments.
+	 * @returns A FromAIArgument object.
+	 */
+	private parseArguments(argsString: string): FromAIArgument {
+		// Split arguments by commas not inside quotes
+		const args: string[] = [];
+		let currentArg = '';
+		let inQuotes = false;
+		let quoteChar = '';
+		let escapeNext = false;
 
-/**
- * Recursively traverses an array to find placeholder values.
- * @param {unknown[]} arr - The array to traverse.
- * @param {string[]} path - The current path in the array.
- * @param {Map<string, string>} results - Map to store found placeholders.
- */
-function traverseArray(arr: unknown[], path: string[], results: Map<string, string>): void {
-	arr.forEach((item, index) => {
-		const currentPath = [...path, index.toString()];
-		const fullPath = currentPath.join('.');
-
-		if (typeof item === 'string' && item.startsWith("{{ '__PLACEHOLDER")) {
-			// Store placeholder values with their full path
-			results.set(encodeDotNotation(fullPath), item);
-		} else if (Array.isArray(item)) {
-			// Recursively traverse nested arrays
-			traverseArray(item, currentPath, results);
-		} else if (typeof item === 'object' && item !== null) {
-			// Recursively traverse nested objects
-			traverseObject(item as NestedObject, currentPath, results);
+		for (let i = 0; i < argsString.length; i++) {
+			const char = argsString[i];
+
+			if (escapeNext) {
+				currentArg += char;
+				escapeNext = false;
+				continue;
+			}
+
+			if (char === '\\') {
+				escapeNext = true;
+				continue;
+			}
+
+			if (char === '"' || char === "'") {
```

Comment:
should we add backticks as quote-chars as well?

Reply:
Updated + added unit tests